### PR TITLE
Added Christofides algorithm

### PR DESCRIPTION
**Pull request by Toptachamann**
Implemented Christofides algorithm suggested in #669 

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        spanningTreeAlgorithm.getSpanningTree().getEdges().forEach(
+                e -> mstAndMatching.addEdge(graph.getEdgeSource(e), graph.getEdgeTarget(e)));
+
+        // find odd degree vertices
+        Set<V> oddDegreeVertices = new HashSet<>();
+        mstAndMatching.vertexSet().forEach(v -> {
+            if ((mstAndMatching.edgesOf(v).size() & 1) == 1) { // if degree is odd
+                oddDegreeVertices.add(v);
+            }
+        });
```
Comment (AlexandruValeanu) [inline jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java]: You can write this as `Set<V> oddDegreeVertices = mstAndMatching.vertexSet().stream().filter(v -> (mstAndMatching.edgesOf(v).size() & 1) == 1).collect(Collectors.toSet());`.

File: jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        GraphPath<V, DefaultEdge> eulerianCycle = eulerianCycleAlgorithm.getEulerianCycle(mstAndMatching);
+
+        // form a closed tour from the Hamiltonian cycle
+        Set<V> visited = new HashSet<>(n);
+        List<V> tourVertices = new ArrayList<>(n);
+        eulerianCycle.getVertexList().forEach(v -> {
+            if (visited.add(v)) {
+                tourVertices.add(v);
+            }
+        });
```
Comment (AlexandruValeanu) [inline jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java]: You also can rewrite this using streams and filters.

File: jgrapht-core/src/test/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSPTest.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
+/*
+ * (C) Copyright 2017-2018, by Timofey Chudakov and Contributors.
```
Comment (AlexandruValeanu) [inline jgrapht-core/src/test/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSPTest.java]: Probably should be 2018-2018.

File: jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+ * The following two observations yield the $3/2$ approximation bound:
+ * <ul>
+ * <li>The cost of every minimum spanning tree is less than or equal to the cost of every Hamiltonian cycle
+ * since after one edge removal every Hamiltonian cycle becomes a spanning tree</li>
+ * <li>Twice the cost of a perfect matching in a graph is less than or equal to the cost of every Hamiltonian
+ * cycle. This follows from the fact that after forming a closed tour using the edges of a perfect matching
+ * the cost of the edges not from the matching is greater than or equal to the cost of the matching edges.</li>
+ * </ul>
+ * <p>
+ * For more details, see <i>Christofides, Nicos. (1976). Worst-Case Analysis of a New Heuristic for the Traveling
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java]: It seems that part of the reference is missing? What's that 10 at the end?

File: jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+ * <p>
+ * The <a href="https://en.wikipedia.org/wiki/Travelling_salesman_problem">travelling salesman problem</a>
+ * (TSP) asks the following question: "Given a list of cities and the distances between each pair of cities,
+ * what is the shortest possible route that visits each city exactly once and returns to the origin city?".
+ * In the metric TSP, the intercity distances satisfy the triangle inequality.
+ * <p>
+ * This is an implementation of the <a href="https://en.wikipedia.org/wiki/Christofides_algorithm">
+ * Christofides algorithm</a>. The algorithms is a $3/2$-approximation assuming that the input graph
+ * satisfies triangle inequality and all edge weights are nonnegative. The implementation requires the input
+ * graph to be undirected and complete. The worst case running time complexity is $\mathcal{O}(V^3E)$.
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java]: I would put undirected and complete in italics. These 2 are quite important

File: jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+     * @throws IllegalArgumentException if the graph is not complete
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
+        if (!graph.getType().isUndirected()) {
+            throw new IllegalArgumentException("Graph must be undirected");
+        }
+        if (!GraphTests.isComplete(graph)) {
+            throw new IllegalArgumentException("Graph is not complete");
```
Comment (jkinable) [inline jgrapht-core/src/main/java/org/jgrapht/alg/tour/ChristofidesThreeHalvesApproxMetricTSP.java]: -> Graph must be complete

**Comment by jkinable** [comment]
Thanks!

**Pull request merged**